Keep fileid when restoring version, using stream copy

PVince81 opened the pull request:
Always stream copy when restoring version

This is a workaround to make it possible to keep the file id when restoring a file. Please note that a stream copy would happen anyway when restoring cross-storage.

Please review @icewind1991 @SergioBertolinSG @nickvergessen @MorrisJobke @DeepDiver1975

Commit:
Fix version rollback to keep fileid

PVince81 commented:
Note that this is broken on OC 8.0.3 as well, we might want a backport @karlitschek (even if it's for 8.0.4, depending on severity)

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/839b8515-bddf-43db-84e1-1397fa3ab007) was created.

PVince81 commented:
Fixes https://github.com/owncloud/core/issues/15468

karlitschek commented:
looks good :+1: Backport is fine but only for 8.0.4

SergioBertolinSG commented:
I've tested #15468 in this branch and it works.
But, if the user2 restores the oldest version, it is restored and a new version (additional) is created.

PVince81 commented:
Yes, this is expected, in case you want to re-switch back to the version you had before.
Can you compare with OC 7 ?

SergioBertolinSG commented:
In OC 7 after restoring last version there is no creation of a new version.
If you have a file with 3 versions, after restoring the oldest i think you should have still 3 versions. OC 7 behaves this way. 
Here after recipient restoring of the oldest you have 4.

PVince81 commented:
Hmmm strange, thanks for the check.
Maybe it's the stream copy that is firing hooks and automatically creating a new version. Not good.

In OC 7 that would probably happen too if the version is restored to external storage... as a side effect

PVince81 commented:
As this is not going into 8.0.3 I'll have a look later during the bugfix phase.

PVince81 commented:
- [ ] BUG: do not create an additional version when restoring version (possibly due to streamcopy), might need a different approach...

nickvergessen commented:
What happens if we don't have enough space to duplicate the file?

PVince81 commented:
The same as on OC 7: stuff will likely break...

Before restoring the version, there is existing code that will backup the current file (as an extra version) through copying, and then proceed to restore the selected version.
This should be addressed (and confirmed) as a separate issue.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/4b907707-cc9b-4145-be84-151da1feacfa) was created.

PVince81 commented:
I changed the implementation: now it's doing the rename/copy on storage level to avoid touching the cache, which will also keep the file id.
The unit tests was also extended to test with single storage or cross storage.
I manually tested restoring versions on single storage and on external storage, the fileid / favorite info / share info was kept.

Please review/retest @icewind1991 @SergioBertolinSG @schiesbn @nickvergessen

PVince81 commented on apps/files_versions/lib/storage.php:
```diff
@@ -295,6 +299,37 @@ public static function rollback($file, $revision) {
+
+		if ($result !== false) {
+			$storage1->unlink($internalPath1);
+			$storage1->getCache()->remove($internalPath1);
```
It turns out it also works without this line. I'll remove it once Jenkins is done (to avoid confusing it)
Maybe getVersions() is doing a rescan internally.

PVince81 commented on apps/files_versions/lib/storage.php:
```diff
@@ -295,6 +299,37 @@ public static function rollback($file, $revision) {
+		$source = $storage1->fopen($internalPath1, 'r');
+		$target = $storage2->fopen($internalPath2, 'w');
+		// FIXME: might need to use part file to avoid concurrent writes
+		// (this would be an issue anyway when renaming/restoring cross-storage)
```
https://github.com/owncloud/core/issues/13756

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/4b907707-cc9b-4145-be84-151da1feacfa) completed: **5 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/b2f381f4-e833-4836-9af8-d1d915002d9e) was created.

PVince81 commented:
Please review (sorry, forgot to set the label to "to review")

PVince81 commented:
@SergioBertolinSG can you retest this ?

SergioBertolinSG commented:
@PVince81 I have tested #15468 in ´´´versions-rollbackkeepfileid´´, here you can restore a shared file to previous version correctly.
But there is still the problem about a version of the file deleted when restored the last version, only in not shared files. Do you remember if there is a different issue about this or this was introduced here?.

PVince81 commented:
@SergioBertolinSG I did fix the problem with the extra version. Did you "git pull" this branch ?
Might be safer to just delete the branch with "git branch -D versions-rollbackkeepfileid", then "git fetch" and then check out the branch again.

SergioBertolinSG commented:
I tested it in a new install, checking out the branch for first time. I have to say it doesn't happen always, but many times.

PVince81 commented:
Are you able to identify the exact cases in which it happens ?
Did it involve external storages ?

SergioBertolinSG commented:
No not external storages. 
Just create five versions of a file and restore the last of the dropdown. (it has just erased me 3 of the 5 versions)

PVince81 commented:
Erasing? That's the expiration code in action. The expiration is only triggered when you do an action on the file. You were talking about additional versions ?

PVince81 commented:
Maybe one way to test it is to look at the "data/$user/files_versions" folder and look at the timestamps "filename.txt.v123456" (the number is the timestamp).
When restoring a version, for example "456789", it should:
1) Make a backup of "filename.txt" by creating a version for it with its current timestamp
2) Copy "filename.txt.v456789" and replace the original "filename.txt".

So at this point, no new version is created.
If other versions unrelated to "456789" or the backup disappear from the list, it's fine. (it's the expiration logic)

SergioBertolinSG commented:
@PVince81 Yes, that's the behaviour. After restoring last version of the dropdown having five versions of a file, in the "data/$user/files_versions" folder there is the backup of the filename.txt, and another version which was the third one. The rest were erased.

So this is the expected behaviour of versions? 
Even if your quota used is close to 0%?

icewind1991 commented on apps/files_versions/lib/storage.php:
```diff
@@ -295,6 +299,36 @@ public static function rollback($file, $revision) {
+		$target = $storage2->fopen($internalPath2, 'w');
+		// FIXME: might need to use part file to avoid concurrent writes
+		// (this would be an issue anyway when renaming/restoring cross-storage)
+		list(, $result) = \OC_Helper::streamCopy($source, $target);
```
use `moveFromStorage` now that's in master

PVince81 replied:
Are you saying I could also get rid of the `$storage1 === $storage2` case that way ?
I'll have a try later.

Commit:
Use moveFromStorage instead of streamCopy when restoring version

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/e5235e2d-a02b-4a26-b4f4-746858bb91cf) was created.

PVince81 commented:
Rebased onto master.

@icewind1991 I replaced streamCopy with moveFromStorage.
Unit tests still work, restoring version still keeps shares in local storage or external storage.

Please rereview @SergioBertolinSG @icewind1991 @schiesbn

SergioBertolinSG commented:
@PVince81 Tested https://github.com/owncloud/core/issues/15468 with last changes. It is working fine :+1:
I think we can discuss that deletion of versions in another issue/irc.

MorrisJobke commented:
@owncloud-bot retest this please

MorrisJobke commented:
Tested and works. File ID keeps stable :+1:

MorrisJobke commented:
Backport in #16019
